receiver/snmpreceiver/scraper.go

```diff
@@ -39,13 +66,359 @@ func newScraper(logger *zap.Logger, cfg *Config, settings component.ReceiverCrea
+
+	// If no column metric configs, nothing else to do
+	if len(metricColumnOIDs) == 0 {
+		return
+	}
+
+	// Retrieve all SNMP indexed data from column metric OIDs
+	indexedData := s.client.GetIndexedData(metricColumnOIDs, scraperErrors)
+	// For each piece of SNMP data, attempt to create the necessary OTEL structures (resources/metrics/datapoints)
+	for _, data := range indexedData {
+		if err := s.indexedDataToMetric(data, metricHelper, configHelper, indexedAttributeValues, indexedResourceAttributeValues); err != nil {
+			scraperErrors.AddPartial(1, fmt.Errorf(errMsgIndexedMetricOIDProcessing, data.oid, data.parentOID, err))
+		}
+	}
+}
+
+// indexedDataToMetric will take one piece of column OID SNMP indexed metric data and turn it
+// into a datapoint for either a new or existing metric with attributes that belongs to either
+// a new or existing resource
+func (s *snmpScraper) indexedDataToMetric(
+	data SNMPData,
+	metricHelper *otelMetricHelper,
+	configHelper *configHelper,
+	indexedAttributeValues map[indexedAttributeKey]string,
+	indexedResourceAttributeValues map[indexedAttributeKey]string,
+) error {
+	// Return an error if this SNMP indexed data is not of a useable type
+	if data.valueType == notSupportedVal || data.valueType == stringVal {
+		return fmt.Errorf(errMsgIndexedBadValueType, data.oid, data.parentOID)
+	}
+
+	// Retrieve the metric config for this SNMP indexed data
+	metricName := configHelper.getMetricName(data.parentOID)
+	metricCfg := configHelper.getMetricConfig(metricName)
+
+	indexString := strings.TrimPrefix(data.oid, data.parentOID)
+
+	// Get data point attributes
+	dataPointAttributes, err := getIndexedDataPointAttributes(configHelper, data.parentOID, indexString, indexedAttributeValues)
+	if err != nil {
+		return fmt.Errorf(errMsgOIDAttributeEmptyValue, metricName, err)
+	}
+
+	// Get resource attributes
+	resourceAttributes, err := getResourceAttributes(configHelper, data.parentOID, indexString, indexedResourceAttributeValues)
+	if err != nil {
+		return fmt.Errorf(errMsgOIDResourceAttributeEmptyValue, metricName, err)
+	}
+
+	// Create a resource key using all of the relevant resource attribute names along
+	// with the row index of the SNMP data
+	resourceAttributeNames := configHelper.getResourceAttributeNames(data.parentOID)
+	resourceKey := getResourceKey(resourceAttributeNames, indexString)
+
+	// Create a new resource if needed
+	resource := metricHelper.getResource(resourceKey)
+	if resource == nil {
+		metricHelper.createResource(resourceKey, resourceAttributes)
+	}
+
+	// Create a new metric if needed
+	metric := metricHelper.getMetric(resourceKey, metricName)
+	if metric == nil {
+		if _, err := metricHelper.createMetric(resourceKey, metricName, metricCfg); err != nil {
+			return err
+		}
+	}
+
+	// Add data point to metric
+	if _, err := metricHelper.addMetricDataPoint(resourceKey, metricName, metricCfg, data, dataPointAttributes); err != nil {
+		return err
+	}
+
+	return nil
+}
```

Comment:
These two functions are also almost exactly the same. Is there a way to unify them?

Reply:
@djaglowski i combined as much like code here as i could into a third method that they both will use, but i don't necessarily think its "better". there is less code (at the expense of some slight efficiency), but i don't know if it's actually more readable.